### PR TITLE
admin: fix broken UI pages when called with empty query params

### LINKED ISSUE #32794: " not found in config dump" when clicking from the admin UI
**If you are reporting *any* crash or *any* potential security issue, *do not*
open an issue in this repo. Please report the issue via emailing
envoy-security@googlegroups.com where the issue will be triaged appropriately.**

*Title*: " not found in config dump" when clicking from the admin UI
*Description*:
If you click the configdump button in the UI you get sent to `/config_dump?resource=&mask=&name_regex=`.
This always displays `not found in config dump`.

Either it should not set the query params, or empty query params should be treated the same as unset I think

*Repro steps*:
> Include sample requests, environment, etc. All data and inputs
required to reproduce the bug.

>**Note**: The [Envoy_collect tool](https://github.com/envoyproxy/envoy/blob/main/tools/envoy_collect/README.md)
gathers a tarball with debug logs, config and the following admin
endpoints: /stats, /clusters and /server_info. Please note if there are
privacy concerns, sanitize the data prior to sharing the tarball/pasting.

*Admin and Stats Output*:
>Include the admin output for the following endpoints: /stats,
/clusters, /routes, /server_info. For more information, refer to the
[admin endpoint documentation.](https://www.envoyproxy.io/docs/envoy/latest/operations/admin)

>**Note**: If there are privacy concerns, sanitize the data prior to
sharing.

*Config*:
>Include the config used to configure Envoy.

*Logs*:
>Include the access logs and the Envoy logs.

>**Note**: If there are privacy concerns, sanitize the data prior to
sharing.

*Call Stack*:
> If the Envoy binary is crashing, a call stack is **required**.
Please refer to the [Bazel Stack trace documentation](https://github.com/envoyproxy/envoy/tree/main/bazel#stack-trace-symbol-resolution).

### PR DESCRIPTION
**Pull request by turettn**
Commit Message:  fix broken UI pages when called with empty query params
Additional Description:

The admin interface handles empty query params all sorts of different ways.  When I was going through and trying to swap out the data structure used for query parameters, I conflated some of the cases.  Looking through the code, I think that this fixes all unintentional behavior changes within the admin application.

Risk Level: Low
Testing:
* Unit tests all work.
* Added a unit test for the helper function
* Manually tested with the standard envoy demo config + an admin server
   * Clicking on the `config_dump` link from the admin landing page works
   * Entering `static_route_configs` in the resource box gives me just that section
   * Entering `bootstrap` in the mask box gives me just that section

Docs Changes: N/A
Release Notes: Bug-fix noted in changelog
Platform Specific Features: N/A

Fixes #32794
Regression introduced in #30964

### CODE REVIEW

**Comment by turettn** [comment]
@jmarantz I don't have permission to add you to the reviewers, but here's the PR [we were talking about](https://github.com/envoyproxy/envoy/pull/30964#issuecomment-1988557771).

**Comment by jmarantz** [comment]
Nice! Were there any other admin handlers you modified to use the new query-param class? If so can you test those also?

WDYT of also adding unit test for the admin-handlers you changed, to provide for it the same URLs that the UI generates?

File: source/server/admin/config_dump_handler.cc [hunk 1]
```diff
@@ -151,8 +151,8 @@ Http::Code ConfigDumpHandler::handlerConfigDump(Http::ResponseHeaderMap& respons
                                                 Buffer::Instance& response,
                                                 AdminStream& admin_stream) const {
   Http::Utility::QueryParamsMulti query_params = admin_stream.queryParams();
-  const auto resource = query_params.getFirstValue("resource");
-  const auto mask = query_params.getFirstValue("mask");
+  const auto resource = Utility::notEmptyQueryParam(query_params, "resource");
```
Comment (jmarantz) [inline source/server/admin/config_dump_handler.cc:154]: pre-existing but I think it's not obvoius what type this returns. Can you use an explicit type here?

The style guide suggests using `auto` only when the type is obvious from context.

File: source/server/admin/utils.cc [hunk 2]
```diff
@@ -46,9 +46,21 @@ absl::Status histogramBucketsParam(const Http::Utility::QueryParamsMulti& params
   return absl::OkStatus();
 }
 
+// Helper method to get a query parameter.
+// Returns the first value for that query parameter, unless that value is empty.
+// In that case, it returns nullopt.
+absl::optional<std::string> notEmptyQueryParam(const Http::Utility::QueryParamsMulti& params,
```
Comment (jmarantz) [inline source/server/admin/utils.cc:52]: nit: nonEmpty?

I'm OK if you decide you prefer notEmpty but to me, "not" suggests this is returning a bool.

**Comment by jmarantz** [comment]
/wait

note you will have to merge main to resolve the conflict on the release notes; you can do that after we are done with the review.

File: source/server/admin/init_dump_handler.cc [hunk 3]
```diff
@@ -13,7 +13,8 @@ InitDumpHandler::InitDumpHandler(Server::Instance& server) : HandlerContextBase(
 Http::Code InitDumpHandler::handlerInitDump(Http::ResponseHeaderMap& response_headers,
                                             Buffer::Instance& response,
                                             AdminStream& admin_stream) const {
-  const auto mask = Utility::notEmptyQueryParam(admin_stream.queryParams(), "mask");
+  const absl::optional<std::string> mask =
+      Utility::nonEmptyQueryParam(admin_stream.queryParams(), "mask");
```
Comment (jmarantz) [inline source/server/admin/init_dump_handler.cc:17]: add a test for this one?

**Review by jmarantz — COMMENTED** [review]
lgtm -- can you add one more unit test?

Can you manually confirm that the all the admin pages all work now from the UI?

**Comment by turettn** [comment]
@jmarantz 

I made the changes you suggested, and merged from main while I was at it.

> Nice! Were there any other admin handlers you modified to use the new query-param class? 

I initially [changed init_dump](https://github.com/envoyproxy/envoy/pull/30964/files#diff-5c6e234322e6b45d8225e7b866e12ea0eed3dc452ffee0e906741882b9d7fabcR16) and this PR attempts to change that back.  

A couple other things also got changed (e.g. [stats](https://github.com/envoyproxy/envoy/pull/30964/files#diff-51e588c59c100fc692c70a163c29d7002804a6b6e10f3e378b76a46510d762caL12)), but included a check to ignore empty query params.

> If so can you test those also?

On my test instance `/init_dump` and `/init_dump?mask=` (where the landing page sends me) both give me an empty response (`{}`).  If you have any hints on how to get something to populate here, I'm happy to try, but it's not part of the admin UI that I've ever touched.

I'll see what I can do about the unit test for that.

As best I can tell, the admin UI works fine in manual testing, but as noted above I have no idea how to get init_dump to populate anything.

**Review by jmarantz — DISMISSED** [review]

**Review by ggreenway — APPROVED** [review]

**Pull request merged**